src/build/external.rs

```diff
@@ -0,0 +1,101 @@
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Performs a build using a provided black-box build command, which ought to
+//! return a list of save-analysis JSON files to be reloaded by the RLS.
+//! Please note that since the command is ran externally (at a file/OS level)
+//! this doesn't work with files that are not saved.
+
+use std::io::Read;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+
+use super::BuildResult;
+
+use rls_data::Analysis;
+use log::{log, trace};
+
+/// Performs a build using an external command and interprets the results.
+/// The command should output on stdout a list of save-analysis .json files
+/// to be reloaded by the RLS.
+/// Note: This is *very* experimental and preliminary - this can viewed as
+/// an experimentation until a more complete solution emerges.
+pub(super) fn build_with_external_cmd<S: AsRef<str>>(path: S, build_dir: PathBuf) -> BuildResult {
+    let path = path.as_ref();
+    let (cmd, args) = {
+        let mut words = path.split_whitespace();
+        let cmd = match words.next() {
+            Some(cmd) => cmd,
+            None => {
+                return BuildResult::Err("Specified build_command is empty".into(), None);
+            }
+        };
+        (cmd, words)
+    };
+    let spawned = Command::new(&cmd)
+        .args(args)
+        .current_dir(&build_dir)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::null())
+        .spawn();
+
+    let child = match spawned {
+        Ok(child) => child,
+        Err(io) => {
+            let err_msg = format!("Couldn't execute: {} ({:?})", path, io.kind());
+            trace!("{}", err_msg);
+            return BuildResult::Err(err_msg, Some(path.to_owned()));
+        },
+    };
+
+    // TODO: Timeout?
+    let reader = std::io::BufReader::new(child.stdout.unwrap());
+    use std::io::BufRead;
+
+    let files = reader.lines().filter_map(|res| res.ok())
+        .map(PathBuf::from)
+        // Relative paths are relative to build command, not RLS itself (cwd may be different)
+        .map(|path| if !path.is_absolute() { build_dir.join(path) } else { path });
+
+    let analyses = match read_analysis_files(files) {
+        Ok(analyses) => analyses,
+        Err(cause) => {
+            let err_msg = format!("Couldn't read analysis data: {}", cause);
+            return BuildResult::Err(err_msg, Some(path.to_owned()));
+        }
+    };
+
+    BuildResult::Success(build_dir.clone(), vec![], analyses, false)
+}
+
+/// Reads and deserializes given save-analysis JSON files into corresponding
+/// `rls_data::Analysis` for each file. If an error is encountered, a `String`
+/// with the error message is returned.
+fn read_analysis_files<I>(files: I) -> Result<Vec<Analysis>, String>
```

Comment:
I think a better approach is to add another `AnalysisLoader` (see https://github.com/nrc/rls-analysis/blob/master/src/loader.rs) implementation to do the loading, that would keep the reading and deserialising of save-analysis encapsulated in rls-analysis rather than both there and the rls.

Reply:
I originally didn't go with that, since it seems you can't incrementally load analysis with a different `AnalysisLoader` - we can only create a fresh one with a different loader.

I imagine we'd have to create new rls-analysis API that would allow us to incrementally load-once with a `&dyn AnalysisLoader` or allow us to replace the `AnalysisHost` with a new one, retaining current index, but with a different generic `AnalysisLoader`?

Reply:
Hmm, that sounds like a problem with the `AnalysisLoader` API, could you file an issue with details to the rls-analysis crate please?